jax/_src/core.py

```diff
@@ -364,21 +364,55 @@ def __repr__(self):
 
 Atom = Union[Var, Literal]
 
+# Table that stores all primitive definitions.
+# Needed so that primitives are treated as singletons
+# when using cloudpickle.
+# Mapping type: (namespace, name) -> Primitive
+_PRIMITIVES_TABLE_: dict[tuple[str, str], Primitive] = {}
+
+def primitives_table_get(namespace: str, name: str):
+  if (namespace, name) in _PRIMITIVES_TABLE_:
+    return _PRIMITIVES_TABLE_[(namespace, name)]
+  raise NotImplementedError(
+    f"Op {(namespace, name)} not found in primitives table. (Did you import it yet?)."
+    f"Primitive table has {list(_PRIMITIVES_TABLE_.keys())}")
+
+def primitives_table_set(namespace: str, name: str, primitive: Primitive):
+  assert (namespace, name) not in _PRIMITIVES_TABLE_, (
+    f"The op name {(namespace, name)} is already taken. "
+    "Try changing the namespace with Primitive(..., namespace='<YOUR_PROJECT_NAME>')."
+  )
+  _PRIMITIVES_TABLE_[(namespace, name)] = primitive
+
+
 class Primitive:
+  # used to identify primitive singletons.
   name: str
+  namespace: str
   # set for multi-output primitives.
   multiple_results: bool = False
   # set for call primitives processed in final style.
   call_primitive: bool = False
   # set for map primitives processed in final style.
   map_primitive: bool = False
 
-  def __init__(self, name: str):
+  def __init__(self, name: str, namespace: str="__jax_core__"):
     self.name = name
+    self.namespace = namespace
+    primitives_table_set(namespace, name, self)
 
   def __repr__(self):
     return f'{self.name}'
 
+  def __reduce_ex__(self, __protocol) -> str | tuple[Any, ...]:
```

Comment:
This is the problematic line, because this is being run on Python 3.9 which does not support `|` syntax for type annotations. That said, I'm a bit confused because this file does `from __future__ import annotations` so I would have thought it would be OK.

Reply:
In any case, I'd try changing this to `Union[str, tuple[Any, ...]]` and then I suspect the doc build will pass.

Reply:
I can just do `tuple[Any, ...]` as I don't return a string ever.

Reply:
> That said, I'm a bit confused because this file does from __future__ import annotations so I would have thought it would be OK.

I also did it in source_info_util which does not import from `__future__`